core/store/src/trie/mem/flexible_data/value.rs

```diff
@@ -0,0 +1,108 @@
+use crate::trie::mem::arena::{ArenaSlice, ArenaSliceMut};
+
+use super::encoding::BorshFixedSize;
+use super::FlexibleDataHeader;
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_primitives::hash::CryptoHash;
+use near_primitives::state::{FlatStateValue, ValueRef};
+
+/// Flexibly-sized data header for a trie value, representing either an inline
+/// value, or a reference to a value stored in the State column.
+///
+/// The flexible part of the data is either the inlined value as a byte array,
+/// or a CryptoHash representing the reference hash.
+#[derive(Clone, Copy, BorshSerialize, BorshDeserialize)]
+pub struct EncodedValueHeader {
+    // The high bit is 1 if the value is inlined, 0 if it is a reference.
+    // The lower bits are the length of the value.
+    length_and_inlined: u32,
```

Comment:
If you’re really looking to minimize memory footprint, this could be an u24, as 23 bits are enough to represent 8MiB, and our current limit on value size is 4MiB. And AFAICT borsh does not need alignment, so this would effectively gain 1 byte.

Not sure it’s worth the hassle, that said.

Reply:
Does u24 exist as a type? I'd have to make a custom borsh serialized type .. yeah..

Reply:
Either making it a custom borsh-serialized type, or implementing it as `low_bits: u16, high_bits: u8` and converting to/from u32 on-the-fly. Not sure which one would be simpler or whether any of the options would be worth the effort, that said.